File: packages/webfetch/src/format.ts
import { appendFileSync, mkdirSync, writeFileSync } from "node:fs";
import path from "node:path";
import { randomUUID } from "node:crypto";
import type { FetchMetadata } from "./types.js";

/**
 * Render the <request>...</request> block that opens every ok /
 * redirect_loop / http_error result. Uniform shape across kinds so the
 * model parses the same surface regardless.
 */
export function renderRequestBlock(meta: FetchMetadata): string {
  const lines = [
    `<request>`,
    `  <url>${meta.url}</url>`,
    `  <final_url>${meta.finalUrl}</final_url>`,
    `  <method>${meta.method}</method>`,
    `  <status>${meta.status}</status>`,
    `  <content_type>${meta.contentType}</content_type>`,
    `  <redirect_chain>${meta.redirectChain.join(" -> ")}</redirect_chain>`,
    `</request>`,
  ];
  return lines.join("\n");
}

export function formatOkText(args: {
  meta: FetchMetadata;
  extractHint: "markdown" | "raw" | "both";
  markdown?: string;
  raw?: string;
  logPath?: string;
  byteCap: boolean;
  totalBytes: number;
}): string {
  const header = renderRequestBlock(args.meta);
  const bodyAttr = args.extractHint;
  let bodyInner: string;
  if (args.extractHint === "markdown" && args.markdown !== undefined) {
    bodyInner = args.markdown;
  } else if (args.extractHint === "raw" && args.raw !== undefined) {
    bodyInner = args.raw;
  } else if (args.extractHint === "both") {
    bodyInner = `<markdown>\n${args.markdown ?? ""}\n</markdown>\n<raw_body>\n${args.raw ?? ""}\n</raw_body>`;
  } else {
    bodyInner = "";
  }
  const bodyBlock = `<body extract="${bodyAttr}">\n${bodyInner}\n</body>`;
  let hint: string;
  if (args.byteCap && args.logPath !== undefined) {
    hint = `(Response exceeded inline cap; showing head+tail of ${args.totalBytes} bytes. Full response at ${args.logPath} — Read with offset/limit to paginate.)`;
  } else {
    const warn =
      args.meta.url !== args.meta.finalUrl &&
      hostOf(args.meta.url) !== hostOf(args.meta.finalUrl)
        ? ` (Final URL host differs from original: ${hostOf(args.meta.url)} -> ${hostOf(args.meta.finalUrl)}. Verify this is expected.)`
        : "";
    const cacheTag =
      args.meta.fromCache && args.meta.cacheAgeSec !== undefined
        ? ` (Served from session cache; age ${args.meta.cacheAgeSec}s.)`
        : "";
    hint = `(Response complete. ${args.totalBytes} bytes total. Content-type: ${args.meta.contentType || "unknown"}. Fetched in ${args.meta.fetchedMs}ms.${warn}${cacheTag})`;
  }
  return [header, bodyBlock, hint].join("\n");
}

export function formatRedirectLoopText(args: {
  meta: FetchMetadata;
  maxRedirects: number;
}): string {
  const header = renderRequestBlock(args.meta);
  const hint = `(Redirect limit (${args.maxRedirects}) exceeded. Chain: ${args.meta.redirectChain.join(" -> ")}. Set max_redirects higher OR pass the final URL directly.)`;
  return [header, hint].join("\n");
}

export function formatHttpErrorText(args: {
  meta: FetchMetadata;
  body: string;
}): string {
  const header = renderRequestBlock(args.meta);
  const bodyBlock = `<body>\n${args.body}\n</body>`;
  const hint = `(HTTP ${args.meta.status}. ${shortReason(args.meta.status)}. Retry or adjust the request per the body.)`;
  return [header, bodyBlock, hint].join("\n");
}

function shortReason(status: number): string {
  if (status === 400) return "Bad Request";
  if (status === 401) return "Unauthorized — check auth headers";
  if (status === 403) return "Forbidden — check permissions or auth";
  if (status === 404) return "Not Found";
  if (status === 408) return "Request Timeout";
  if (status === 410) return "Gone";
  if (status === 418) return "I'm a teapot";
  if (status === 429) return "Too Many Requests — back off";
  if (status === 500) return "Internal Server Error";
  if (status === 502) return "Bad Gateway";
  if (status === 503) return "Service Unavailable";
  if (status === 504) return "Gateway Timeout";
  if (status >= 400 && status < 500) return "Client error";
  if (status >= 500) return "Server error";
  return "Non-success status";
}

export function hostOf(url: string): string {
  try {
    return new URL(url).hostname;
  } catch {
    return "";
  }
}

// ---- spill-to-file ----

/**
 * Save the full response body to a spill file under
 * ~/.agent-sh/webfetch-cache/<session>/<uuid>.<ext>. Returns the path.
 */
export function spillToFile(args: {
  bytes: Uint8Array;
  dir: string;
  sessionId: string;
  contentType: string;
}): string {
  const dir = path.join(args.dir, args.sessionId);
  mkdirSync(dir, { recursive: true });
  const ext = extensionFor(args.contentType);
  const filePath = path.join(dir, `${randomUUID()}.${ext}`);
  writeFileSync(filePath, Buffer.from(args.bytes));
  return filePath;
}

function extensionFor(contentType: string): string {
  const lower = contentType.toLowerCase();
  if (lower.includes("text/html") || lower.includes("xhtml")) return "html";
  if (lower.includes("json")) return "json";
  if (lower.includes("xml")) return "xml";
  if (lower.includes("csv")) return "csv";
  if (lower.includes("markdown")) return "md";
  if (lower.includes("text/")) return "txt";
  return "bin";
}

/**
 * Append-only version used by future streaming implementations. Kept
 * here for symmetry with bash's stream-to-file pattern; not used by
 * the current engine (which buffers fully before deciding to spill).
 */
export function appendSpill(filePath: string, bytes: Uint8Array): void {
  appendFileSync(filePath, Buffer.from(bytes));
}

/**
 * Given a big body, return head (first N bytes) + tail (last N bytes)
 * text concatenated with an elision marker. Used for the inline view
 * when spilled.
 */
export function headAndTail(
  bytes: Uint8Array,
  headBytes: number,
  tailBytes: number,
  logPath: string,
): string {
  if (bytes.length <= headBytes + tailBytes) {
    return Buffer.from(bytes).toString("utf8");
  }
  const head = Buffer.from(bytes.subarray(0, headBytes)).toString("utf8");
  const tail = Buffer.from(
    bytes.subarray(bytes.length - tailBytes),
  ).toString("utf8");
  const elided = bytes.length - headBytes - tailBytes;
  return `${head}\n\n... (${elided} bytes elided; full response at ${logPath}) ...\n\n${tail}`;
}

File: crates/webfetch/src/format.rs
use std::fs;
use std::path::{Path, PathBuf};
use url::Url;

use crate::types::FetchMetadata;

pub fn render_request_block(meta: &FetchMetadata) -> String {
    let chain = meta.redirect_chain.join(" -> ");
    format!(
        "<request>\n  <url>{}</url>\n  <final_url>{}</final_url>\n  <method>{}</method>\n  <status>{}</status>\n  <content_type>{}</content_type>\n  <redirect_chain>{}</redirect_chain>\n</request>",
        meta.url,
        meta.final_url,
        meta.method.as_str(),
        meta.status,
        meta.content_type,
        chain,
    )
}

pub struct FormatOkArgs<'a> {
    pub meta: &'a FetchMetadata,
    pub extract_hint: &'a str,
    pub markdown: Option<&'a str>,
    pub raw: Option<&'a str>,
    pub log_path: Option<&'a str>,
    pub byte_cap: bool,
    pub total_bytes: usize,
}

pub fn format_ok_text(args: FormatOkArgs<'_>) -> String {
    let header = render_request_block(args.meta);
    let body_inner = match args.extract_hint {
        "markdown" => args.markdown.unwrap_or("").to_string(),
        "raw" => args.raw.unwrap_or("").to_string(),
        "both" => format!(
            "<markdown>\n{}\n</markdown>\n<raw_body>\n{}\n</raw_body>",
            args.markdown.unwrap_or(""),
            args.raw.unwrap_or(""),
        ),
        _ => String::new(),
    };
    let body_block = format!("<body extract=\"{}\">\n{}\n</body>", args.extract_hint, body_inner);

    let hint = if args.byte_cap && args.log_path.is_some() {
        format!(
            "(Response exceeded inline cap; showing head+tail of {} bytes. Full response at {} — Read with offset/limit to paginate.)",
            args.total_bytes,
            args.log_path.unwrap(),
        )
    } else {
        let original_host = host_of(&args.meta.url);
        let final_host = host_of(&args.meta.final_url);
        let warn = if args.meta.url != args.meta.final_url && original_host != final_host {
            format!(
                " (Final URL host differs from original: {} -> {}. Verify this is expected.)",
                original_host, final_host
            )
        } else {
            String::new()
        };
        let cache_tag = if args.meta.from_cache {
            let age = args.meta.cache_age_sec.unwrap_or(0);
            format!(" (Served from session cache; age {}s.)", age)
        } else {
            String::new()
        };
        let ct = if args.meta.content_type.is_empty() {
            "unknown".to_string()
        } else {
            args.meta.content_type.clone()
        };
        format!(
            "(Response complete. {} bytes total. Content-type: {}. Fetched in {}ms.{}{})",
            args.total_bytes, ct, args.meta.fetched_ms, warn, cache_tag
        )
    };

    format!("{}\n{}\n{}", header, body_block, hint)
}

pub struct FormatRedirectLoopArgs<'a> {
    pub meta: &'a FetchMetadata,
    pub max_redirects: u32,
}

pub fn format_redirect_loop_text(args: FormatRedirectLoopArgs<'_>) -> String {
    let header = render_request_block(args.meta);
    let chain = args.meta.redirect_chain.join(" -> ");
    let hint = format!(
        "(Redirect limit ({}) exceeded. Chain: {}. Set max_redirects higher OR pass the final URL directly.)",
        args.max_redirects, chain
    );
    format!("{}\n{}", header, hint)
}

pub struct FormatHttpErrorArgs<'a> {
    pub meta: &'a FetchMetadata,
    pub body: &'a str,
}

pub fn format_http_error_text(args: FormatHttpErrorArgs<'_>) -> String {
    let header = render_request_block(args.meta);
    let body_block = format!("<body>\n{}\n</body>", args.body);
    let hint = format!(
        "(HTTP {}. {}. Retry or adjust the request per the body.)",
        args.meta.status,
        short_reason(args.meta.status),
    );
    format!("{}\n{}\n{}", header, body_block, hint)
}

fn short_reason(status: u16) -> &'static str {
    match status {
        400 => "Bad Request",
        401 => "Unauthorized — check auth headers",
        403 => "Forbidden — check permissions or auth",
        404 => "Not Found",
        408 => "Request Timeout",
        410 => "Gone",
        418 => "I'm a teapot",
        429 => "Too Many Requests — back off",
        500 => "Internal Server Error",
        502 => "Bad Gateway",
        503 => "Service Unavailable",
        504 => "Gateway Timeout",
        s if (400..500).contains(&s) => "Client error",
        s if s >= 500 => "Server error",
        _ => "Non-success status",
    }
}

pub fn host_of(url: &str) -> String {
    Url::parse(url)
        .ok()
        .and_then(|u| u.host_str().map(|s| s.to_string()))
        .unwrap_or_default()
}

// ---- spill-to-file ----

pub struct SpillArgs<'a> {
    pub bytes: &'a [u8],
    pub dir: &'a Path,
    pub session_id: &'a str,
    pub content_type: &'a str,
}

pub fn spill_to_file(args: SpillArgs<'_>) -> std::io::Result<PathBuf> {
    let dir = args.dir.join(args.session_id);
    fs::create_dir_all(&dir)?;
    let ext = extension_for(args.content_type);
    let filename = format!("{}.{}", uuid::Uuid::new_v4(), ext);
    let path = dir.join(filename);
    fs::write(&path, args.bytes)?;
    Ok(path)
}

fn extension_for(content_type: &str) -> &'static str {
    let lower = content_type.to_ascii_lowercase();
    if lower.contains("text/html") || lower.contains("xhtml") {
        "html"
    } else if lower.contains("json") {
        "json"
    } else if lower.contains("xml") {
        "xml"
    } else if lower.contains("csv") {
        "csv"
    } else if lower.contains("markdown") {
        "md"
    } else if lower.contains("text/") {
        "txt"
    } else {
        "bin"
    }
}

/// Return head (first N bytes) + tail (last N bytes) concatenated with
/// an elision marker. Mirrors the bash head+tail spill pattern.
pub fn head_and_tail(
    bytes: &[u8],
    head_bytes: usize,
    tail_bytes: usize,
    log_path: &str,
) -> String {
    if bytes.len() <= head_bytes + tail_bytes {
        return String::from_utf8_lossy(bytes).into_owned();
    }
    let head = String::from_utf8_lossy(&bytes[..head_bytes]).into_owned();
    let tail = String::from_utf8_lossy(&bytes[bytes.len() - tail_bytes..]).into_owned();
    let elided = bytes.len() - head_bytes - tail_bytes;
    format!(
        "{}\n\n... ({} bytes elided; full response at {}) ...\n\n{}",
        head, elided, log_path, tail
    )
}
